File: PRNumberBiz/app/src/main/java/com/pplus/prnumberbiz/apps/setting/ui/FAQActivity.java
package com.pplus.prnumberbiz.apps.setting.ui;

import android.content.Intent;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.view.View;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.pplus.prnumberbiz.Const;
import com.pplus.prnumberbiz.R;
import com.pplus.prnumberbiz.apps.common.toolbar.ImplToolbar;
import com.pplus.prnumberbiz.apps.common.toolbar.OnToolbarListener;
import com.pplus.prnumberbiz.apps.common.toolbar.ToolbarOption;
import com.pplus.prnumberbiz.apps.common.ui.base.BaseActivity;
import com.pplus.prnumberbiz.apps.setting.data.FaqAdapter;
import com.pplus.prnumberbiz.core.network.ApiBuilder;
import com.pplus.prnumberbiz.core.network.model.dto.Faq;
import com.pplus.prnumberbiz.core.network.model.dto.FaqGroup;
import com.pplus.prnumberbiz.core.network.model.response.NewResultResponse;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import network.common.PplusCallback;
import retrofit2.Call;

public class FAQActivity extends BaseActivity implements ImplToolbar{

    @Override
    public String getPID(){

        return null;
    }

    @Override
    public int getLayoutRes(){

        return R.layout.activity_faq;
    }

    private List<FaqGroup> groupList;
    private LinearLayout layout_group;
    private ImageView image_faq_down;
    private boolean mIsGroupClick = false;
    private FaqGroup mSelectGroup;
    private TextView text_group;

    private LinearLayoutManager mLayoutManager;
    private FaqAdapter mAdapter;
    private boolean mLockListView = true;
    private int mTotalCount = 0, mPage = 0;

    @Override
    public void initializeView(Bundle savedInstanceState){
        findViewById(R.id.fv_faq).setOnClickListener(categoryLayoutClickListener);
        findViewById(R.id.fv_faq).setEnabled(false);
        layout_group = (LinearLayout)findViewById(R.id.layout_faq_group);
        layout_group.setVisibility(View.GONE);
        image_faq_down =(ImageView)findViewById(R.id.image_faq_down);
        text_group = (TextView)findViewById(R.id.text_faq_group);

        RecyclerView recyclerView = (RecyclerView) findViewById(R.id.recycler_faq);
        mAdapter = new FaqAdapter(this);
        mLayoutManager = new LinearLayoutManager(this);
        mLayoutManager.setOrientation(LinearLayoutManager.VERTICAL);
        recyclerView.setLayoutManager(mLayoutManager);
        recyclerView.setAdapter(mAdapter);

        mAdapter.setOnItemClickListener(new FaqAdapter.OnItemClickListener(){

            @Override
            public void onItemClick(int position){

                if(mAdapter.getItem(position) != null) {
                    // Webview 상세로 이동
                    Intent intent = new Intent(FAQActivity.this, NoticeDetailActivity.class);
                    intent.putExtra(Const.FAQ, mAdapter.getItem(position));
                    startActivity(intent);
                }
            }
        });

        recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener(){

            int pastVisibleItems, visibleItemCount, totalItemCount;

            @Override
            public void onScrolled(RecyclerView recyclerView, int dx, int dy){

                super.onScrolled(recyclerView, dx, dy);
                visibleItemCount = mLayoutManager.getChildCount();
                totalItemCount = mLayoutManager.getItemCount();
                pastVisibleItems = mLayoutManager.findFirstVisibleItemPosition();
                if(!mLockListView) {
                    if((totalItemCount < mTotalCount) && (visibleItemCount + pastVisibleItems) >= (totalItemCount * Const.NEXT_RECYCLERVIEW_MARGIN)) {
                        mPage++;
                        listCall(mPage);
                    }
                }
            }
        });

        groupList = new ArrayList<FaqGroup>();
        FaqGroup totalGroup = new FaqGroup();
        totalGroup.setNo(0l);
        totalGroup.setName(getString(R.string.word_total));
        mSelectGroup = totalGroup;
        groupList.add(totalGroup);
        getCount();
        callGroup();
    }

    public View.OnClickListener categoryLayoutClickListener = new View.OnClickListener(){

        @Override
        public void onClick(View view){

            mIsGroupClick = !mIsGroupClick;

            if(mIsGroupClick) {
                image_faq_down.setImageResource(R.drawable.ic_setting_up);
                layout_group.removeAllViews();
                layout_group.setVisibility(View.VISIBLE);

                for(int i = 0; i < groupList.size(); i++) {
                    if(!groupList.get(i).getNo().equals(mSelectGroup.getNo())){
                        View v = getLayoutInflater().inflate(R.layout.item_toolbar_list_row, null);
                        TextView tv = (TextView) v.findViewById(R.id.tv_item_list);
                        tv.setText(groupList.get(i).getName());
                        tv.setTag(tv.getId(), groupList.get(i));
                        tv.setOnClickListener(categoryClickListener);
                        layout_group.addView(v);
                    }

                }
            } else {
                image_faq_down.setImageResource(R.drawable.ic_setting_down);
                layout_group.setVisibility(View.GONE);
            }
        }
    };

    public View.OnClickListener categoryClickListener = new View.OnClickListener(){

        @Override
        public void onClick(View view){

            mIsGroupClick = false;
            image_faq_down.setImageResource(R.drawable.ic_setting_down);
            layout_group.setVisibility(View.GONE);
            mSelectGroup = (FaqGroup) view.getTag(view.getId());
            text_group.setText(mSelectGroup.getName());
            getCount();
        }
    };

    private void callGroup(){

        Map<String, String> params = new HashMap<>();
        params.put("appKey", getPackageName());
        showProgress("");
        ApiBuilder.create().getFaqGroupAll(params).setCallback(new PplusCallback<NewResultResponse<FaqGroup>>(){

            @Override
            public void onResponse(Call<NewResultResponse<FaqGroup>> call, NewResultResponse<FaqGroup> response){
                hideProgress();
                groupList.addAll(response.getDatas());
                findViewById(R.id.fv_faq).setEnabled(true);
            }

            @Override
            public void onFailure(Call<NewResultResponse<FaqGroup>> call, Throwable t, NewResultResponse<FaqGroup> response){
                hideProgress();
            }
        }).build().call();
    }

    private void getCount(){

        Map<String, String> params = new HashMap<>();
        params.put("appKey", getPackageName());
        if(!mSelectGroup.getNo().equals(0l)){
            params.put("no", "" + mSelectGroup.getNo());
        }
        ApiBuilder.create().getFaqCount(params).setCallback(new PplusCallback<NewResultResponse<Integer>>(){

            @Override
            public void onResponse(Call<NewResultResponse<Integer>> call, NewResultResponse<Integer> response){

                mTotalCount = response.getData();
                mPage = 1;
                mAdapter.clear();
                listCall(mPage);
            }

            @Override
            public void onFailure(Call<NewResultResponse<Integer>> call, Throwable t, NewResultResponse<Integer> response){

            }
        }).build().call();
    }

    private void listCall(int page){
        Map<String, String> params = new HashMap<>();
        params.put("appKey", getPackageName());
        params.put("pg", "" + page);
        if(!mSelectGroup.getNo().equals(0l)){
            params.put("no", "" + mSelectGroup.getNo());
        }
        mLockListView = false;
        showProgress("");
        ApiBuilder.create().getFaqList(params).setCallback(new PplusCallback<NewResultResponse<Faq>>(){

            @Override
            public void onResponse(Call<NewResultResponse<Faq>> call, NewResultResponse<Faq> response){
                hideProgress();
                mLockListView = true;
                mAdapter.addAll(response.getDatas());
            }

            @Override
            public void onFailure(Call<NewResultResponse<Faq>> call, Throwable t, NewResultResponse<Faq> response){
                hideProgress();
            }
        }).build().call();
    }

    @NonNull
    @Override
    public ToolbarOption getToolbarOption(){

        ToolbarOption toolbarOption = new ToolbarOption(this);
        toolbarOption.initializeDefaultToolbar(getString(R.string.word_faq_en), ToolbarOption.ToolbarMenu.LEFT);
        return toolbarOption;
    }

    @Override
    public OnToolbarListener getOnToolbarClickListener(){

        return new OnToolbarListener(){

            @Override
            public void onClick(View v, ToolbarOption.ToolbarMenu toolbarMenu, Object tag){

                switch (toolbarMenu) {
                    case LEFT:
                        if(tag.equals(1)) {
                            onBackPressed();
                        }
                        break;
                }
            }
        };
    }
}

File: PRNumber/app/src/main/java/com/pplus/prnumberuser/apps/setting/ui/FAQActivity.kt
package com.pplus.prnumberuser.apps.setting.ui

import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.pplus.networks.common.PplusCallback
import com.pplus.prnumberuser.Const
import com.pplus.prnumberuser.R
import com.pplus.prnumberuser.apps.common.toolbar.ImplToolbar
import com.pplus.prnumberuser.apps.common.toolbar.OnToolbarListener
import com.pplus.prnumberuser.apps.common.toolbar.ToolbarOption
import com.pplus.prnumberuser.apps.common.toolbar.ToolbarOption.ToolbarMenu
import com.pplus.prnumberuser.apps.common.ui.base.BaseActivity
import com.pplus.prnumberuser.apps.setting.data.FaqAdapter
import com.pplus.prnumberuser.core.network.ApiBuilder
import com.pplus.prnumberuser.core.network.model.dto.Faq
import com.pplus.prnumberuser.core.network.model.dto.FaqGroup
import com.pplus.prnumberuser.core.network.model.response.NewResultResponse
import com.pplus.prnumberuser.databinding.ActivityFaqBinding
import com.pplus.prnumberuser.databinding.ItemToolbarListRowBinding
import retrofit2.Call
import java.util.*

class FAQActivity : BaseActivity(), ImplToolbar {
    override fun getPID(): String? {
        return ""
    }

    private lateinit var binding: ActivityFaqBinding

    override fun getLayoutView(): View {
        binding = ActivityFaqBinding.inflate(layoutInflater)
        return binding.root
    }

    private var groupList: MutableList<FaqGroup>? = null
    private var mIsGroupClick = false
    private var mSelectGroup: FaqGroup? = null
    private var mLayoutManager: LinearLayoutManager? = null
    private var mAdapter: FaqAdapter? = null
    private var mLockListView = true
    private var mTotalCount = 0
    private var mPage = 0

    override fun initializeView(savedInstanceState: Bundle?) {
        binding.fvFaq.setOnClickListener(categoryLayoutClickListener)
        binding.fvFaq.isEnabled = false
        binding.layoutFaqGroup.visibility = View.GONE
        binding.scrollFaqGroup.visibility = View.GONE
        mAdapter = FaqAdapter(this)
        mLayoutManager = LinearLayoutManager(this)
        binding.recyclerFaq.layoutManager = mLayoutManager
        binding.recyclerFaq.adapter = mAdapter
        mAdapter!!.setOnItemClickListener { position ->
            if (mAdapter!!.getItem(position) != null) {
                // Webview 상세로 이동
                val intent = Intent(this@FAQActivity, NoticeDetailActivity::class.java)
                intent.putExtra(Const.FAQ, mAdapter!!.getItem(position))
                startActivity(intent)
            }
        }
        binding.recyclerFaq.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            var pastVisibleItems = 0
            var visibleItemCount = 0
            var totalItemCount = 0
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                visibleItemCount = mLayoutManager!!.childCount
                totalItemCount = mLayoutManager!!.itemCount
                pastVisibleItems = mLayoutManager!!.findFirstVisibleItemPosition()
                if (!mLockListView) {
                    if (totalItemCount < mTotalCount && visibleItemCount + pastVisibleItems >= totalItemCount * Const.NEXT_RECYCLERVIEW_MARGIN) {
                        mPage++
                        listCall(mPage)
                    }
                }
            }
        })
        groupList = ArrayList()
        val totalGroup = FaqGroup()
        totalGroup.no = 0L
        totalGroup.name = getString(R.string.word_total)
        mSelectGroup = totalGroup
        groupList!!.add(totalGroup)
        getCount()
        callGroup()
    }

    var categoryLayoutClickListener = View.OnClickListener {
        mIsGroupClick = !mIsGroupClick
        if (mIsGroupClick) {
            binding.imageFaqDown.setImageResource(R.drawable.ic_setting_up)
            binding.layoutFaqGroup.removeAllViews()
            binding.layoutFaqGroup.visibility = View.VISIBLE
            binding.scrollFaqGroup.visibility = View.VISIBLE
            for (i in groupList!!.indices) {
                if (groupList!![i].no != mSelectGroup!!.no) {
                    val toolbarBinding = ItemToolbarListRowBinding.inflate(layoutInflater)
                    toolbarBinding.tvItemList.text = groupList!![i].name
                    toolbarBinding.tvItemList.setTag(toolbarBinding.tvItemList.id, groupList!![i])
                    toolbarBinding.tvItemList.setOnClickListener(categoryClickListener)
                    binding.layoutFaqGroup.addView(toolbarBinding.root)
                }
            }
        } else {
            binding.imageFaqDown.setImageResource(R.drawable.ic_setting_down)
            binding.layoutFaqGroup.visibility = View.GONE
            binding.scrollFaqGroup.visibility = View.GONE
        }
    }
    var categoryClickListener = View.OnClickListener { view ->
        mIsGroupClick = false
        binding.imageFaqDown.setImageResource(R.drawable.ic_setting_down)
        binding.layoutFaqGroup.visibility = View.GONE
        binding.scrollFaqGroup.visibility = View.GONE
        mSelectGroup = view.getTag(view.id) as FaqGroup
        binding.textFaqGroup.text = mSelectGroup!!.name
        getCount()
    }

    private fun callGroup() {
        val params = HashMap<String, String>()
        params["appKey"] = packageName
        showProgress("")
        ApiBuilder.create().getFaqGroupAll(params).setCallback(object : PplusCallback<NewResultResponse<FaqGroup>> {
            override fun onResponse(call: Call<NewResultResponse<FaqGroup>>?,
                                    response: NewResultResponse<FaqGroup>?) {
                hideProgress()
                if (response?.datas != null) {
                    groupList!!.addAll(response.datas as MutableList)
                }
                binding.fvFaq.isEnabled = true
            }

            override fun onFailure(call: Call<NewResultResponse<FaqGroup>>?,
                                   t: Throwable?,
                                   response: NewResultResponse<FaqGroup>?) {
                hideProgress()
            }

        }).build().call()
    }

    private fun getCount() {
        val params = HashMap<String, String>()
        params["appType"] = Const.APP_TYPE
        if (mSelectGroup!!.no != 0L) {
            params["no"] = "" + mSelectGroup!!.no
        }
        params["platform"] = "aos"
        ApiBuilder.create().getFaqCount(params).setCallback(object : PplusCallback<NewResultResponse<Int>> {

            override fun onResponse(call: Call<NewResultResponse<Int>>,
                                    response: NewResultResponse<Int>) {
                if (response.data != null) {
                    mTotalCount = response.data!!
                    mPage = 1
                    mAdapter!!.clear()
                    listCall(mPage)
                }
            }

            override fun onFailure(call: Call<NewResultResponse<Int>>,
                                   t: Throwable?,
                                   response: NewResultResponse<Int>) {
                hideProgress()
            }
        }).build().call()
    }

    private fun listCall(page: Int) {
        val params = HashMap<String, String>()
        params["appType"] = Const.APP_TYPE
        params["pg"] = "" + page
        if (mSelectGroup!!.no != 0L) {
            params["no"] = "" + mSelectGroup!!.no
        }
        params["platform"] = "aos"
        mLockListView = true
        showProgress("")
        ApiBuilder.create().getFaqList(params).setCallback(object : PplusCallback<NewResultResponse<Faq>> {
            override fun onResponse(call: Call<NewResultResponse<Faq>>, response: NewResultResponse<Faq>) {
                hideProgress()
                if(response.datas != null){
                    mLockListView = false
                    mAdapter!!.addAll(response.datas)
                }
            }

            override fun onFailure(call: Call<NewResultResponse<Faq>>, t: Throwable, response: NewResultResponse<Faq>) {
                hideProgress()
            }
        }).build().call()
    }

    override fun getToolbarOption(): ToolbarOption {
        val toolbarOption = ToolbarOption(this)
        toolbarOption.initializeDefaultToolbar(getString(R.string.word_faq_en), ToolbarMenu.LEFT)
        return toolbarOption
    }

    override fun getOnToolbarClickListener(): OnToolbarListener {
        return object : OnToolbarListener {
            override fun onClick(v: View?, toolbarMenu: ToolbarMenu?, tag: Any?) {
                when (toolbarMenu) {
                    ToolbarMenu.LEFT -> if (tag == 1) {
                        onBackPressed()
                    }
                }
            }
        }
    }
}